crates/holochain_sqlite/src/table.rs

```diff
@@ -0,0 +1,322 @@
+    /// mutable state for the entire Conductor
+    ConductorState,
+    /// database that stores wasm bytecode
+    Wasm,
+    /// database to store the [DnaDef]
+    DnaDef,
+    /// database to store the [EntryDef] Kvv store
+    EntryDef,
+    /// Authored [DhtOp]s KV store
+    AuthoredDhtOps,
+    /// Integrated [DhtOp]s KV store
+    IntegratedDhtOps,
+    /// Integration Queue of [DhtOp]s KV store where key is [DhtOpHash]
+    IntegrationLimbo,
+    /// Place for [DhtOp]s waiting to be validated to hang out. KV store where key is a [DhtOpHash]
+    ValidationLimbo,
+    /// KVV store to accumulate validation receipts for a published EntryHash
+    ValidationReceipts,
+    /// Single store for all known agents on the network
+    Agent,
+
+    #[cfg(feature = "test_utils")]
+    TestSingle(String),
+
+    #[cfg(feature = "test_utils")]
+    TestMulti(String),
+}
+
+impl TableName {
+    /// Associates a [TableKind] to each [TableName]
+    pub fn kind(&self) -> TableKind {
+        use TableKind::*;
+        use TableName::*;
+        match self {
+            ElementVaultPublicEntries => Single,
+            ElementVaultPrivateEntries => Single,
+            ElementVaultHeaders => Single,
+            MetaVaultSys => Multi,
+            MetaVaultLinks => Single,
+            MetaVaultMisc => Single,
+            ChainSequence => Single, // int
+            ElementCacheEntries => Single,
+            ElementCacheHeaders => Single,
+            MetaCacheSys => Multi,
+            MetaCacheLinks => Single,
+            MetaCacheStatus => Single,
+            ConductorState => Single,
+            Wasm => Single,
+            DnaDef => Single,
+            EntryDef => Single,
+            AuthoredDhtOps => Single,
+            IntegratedDhtOps => Single,
+            IntegrationLimbo => Single,
+            ValidationLimbo => Single,
+            ValidationReceipts => Multi,
+            Agent => Single,
+            #[cfg(feature = "test_utils")]
+            TestSingle(_) => Single,
+            #[cfg(feature = "test_utils")]
+            TestMulti(_) => Multi,
+        }
+    }
+}
+
+impl ToSql for TableName {
+    fn to_sql(&self) -> Result<rusqlite::types::ToSqlOutput<'_>> {
+        Ok(rusqlite::types::ToSqlOutput::Owned(
+            format!("{}", self).into(),
+        ))
+    }
+}
+
+pub enum TableKind {
+    Single,
+    Multi,
+}
+
+impl TableKind {
+    pub fn is_single(&self) -> bool {
+        matches!(self, Self::Single)
+    }
+
+    pub fn is_multi(&self) -> bool {
+        matches!(self, Self::Multi)
+    }
+}
+
+pub(crate) fn initialize_table_single(
+    conn: &mut Connection,
+    table_name: String,
+) -> rusqlite::Result<()> {
+    // create table
+    conn.execute(
+        &format!(
+            "CREATE TABLE IF NOT EXISTS {} (
+                key       BLOB PRIMARY KEY,
+                val       BLOB NOT NULL
+            );",
+            table_name
+        ),
+        NO_PARAMS,
+    )?;
```

Comment:
Avoid a format, use sqlite's replacement:

```suggestion
    conn.execute(
        "CREATE TABLE IF NOT EXISTS ?1 (
            key       BLOB PRIMARY KEY,
            val       BLOB NOT NULL
        );",
        params![table_name],
    )?;
```

Reply:
@neonphog I thought so too, but in SQL you can't inject the table name, only things like table fields.